src/tds/codec/pre_login.rs

```diff
@@ -113,19 +134,69 @@ impl Decode<BytesMut> for PreloginMessage {
             // TODO: support parsing more
             match token {
                 // version
-                0 => {
+                0x00 => {
                     ret.version = cursor.read_u32::<BigEndian>()?;
                     ret.sub_build = cursor.read_u16::<BigEndian>()?;
                 }
                 // encryption
-                1 => {
+                0x01 => {
                     let encrypt = cursor.read_u8()?;
                     ret.encryption = tds::EncryptionLevel::try_from(encrypt).map_err(|_| {
                         Error::Protocol(format!("invalid encryption value: {}", encrypt).into())
                     })?;
                 }
-                3 => debug_assert_eq!(length, 0), // threadid
-                4 => debug_assert_eq!(length, 1), // mars
+                // instance name
+                0x02 => {
+                    let mut bytes = Vec::new();
+                    let mut next_byte = cursor.read_u8()?;
+
+                    while next_byte != 0x00 {
+                        bytes.push(next_byte);
+                        next_byte = cursor.read_u8()?;
+
+                        continue;
+                    }
+
+                    if !bytes.is_empty() {
+                        ret.instance_name = Some(String::from_utf8_lossy(&bytes).into_owned());
+                    }
+                }
+                // threadid (should be empty when sent from server to client)
+                0x03 => debug_assert_eq!(length, 0),
+                // mars
+                0x04 => {
+                    ret.mars = cursor.read_u8()? != 0;
+                }
+                // activity id
+                0x05 => {
+                    let mut data = [0u8; 16];
+
+                    for item in data.iter_mut() {
+                        *item = cursor.read_u8()?;
+                    }
+
+                    // Microsoft UUID comes in a "wrong" order.
+                    reorder_bytes(&mut data);
+
+                    ret.activity_id = Some(ActivityId {
+                        id: Uuid::from_bytes(data),
+                        sequence: cursor.read_u32::<LittleEndian>()?,
+                    });
+                }
+                // fed auth
+                0x06 => {
+                    ret.fed_auth_required = cursor.read_u8()? != 0;
+                }
+                // nonce
+                0x07 => {
+                    let mut data = [0u8; 32];
+
+                    for item in data.iter_mut() {
+                        *item = cursor.read_u8()?;
```

Comment:
it looks a bit weird, isn't there a `read()` method to fill the buffer?

Reply:
true.